lib/goldiloader/active_record_patches.rb

```diff
@@ -47,7 +47,7 @@ def auto_include!(auto_include = true)
     end
 
     def auto_include_value
-      @values.fetch(:auto_include, true)
+      @values.fetch(:auto_include) { Goldiloader.auto_include }
```

Comment:
🤔  Just to make sure I understand the intent here: `auto_include` can be set at three different levels with the following order of precedence:

1. `ActiveRecord::Relation` instances either via regular ActiveRecord query scope or an association's scope
2. Within a thread
3. Globally (this wasn't actually implemented but I assume that was the intent based on the README change)

I believe the main scenarios that 2 and 3 are attempting to support (as described in #55) are:

* Goldiloader enabled by default but disabled for a block of code
* Goldiloader disabled by default but enabled for a block of code

This seems like it works reasonably well if `ActiveRecord::Relation` instances only explicitly set `auto_include = false` when automatic eager loading is known to be problematic. However, I'm concerned it could lead to unexpected results if `ActiveRecord::Relation` instances set `auto_include = true` since the developers intent as to how this code should behave in those main scenarios is less clear.

All of this makes me wonder if this solution is too complex. Perhaps a simpler alternative would be to just support a global configuration option which is the default value for `ActiveRecord::Relation#auto_include`? This would allow developers to incrementally rollout automatic eager loading association by association rather than code block by code block which is likely a better adoption strategy.

Reply:
Hi! Before going through with more changes, I think it's best to get everyone aligned on what to do. This seems the best place to do that.

> Just to make sure I understand the intent here:

Personally, I only care about (dynamic) global disable/enable. This needs to be thread-aware, otherwise multi-threaded servers like puma would have the setting enabled/disabled by other threads.

Why do I only care about global, dynamic enabled/disable? Basically, I want to have it disabled by default, and enable it per request/per job when appropriate. The reasons for opt-in instead of opt-out are that some object graphs we deal with can cause extensive memory when preloading everything. Thus, we enable this on select requests/jobs (which is also how I read the intent expressed on https://github.com/salsify/goldiloader/issues/55#issue-230046390).

This PR (attempts to) preserve the `auto_include` handling, as requested by https://github.com/salsify/goldiloader/issues/55#issuecomment-427937927 .

I explicitly did not add "default enabled, disable per block" because I don't really find that scenario particularly useful, and keeps both the code and the API simpler.

Does this make sense? And does it make the code make more sense?

Reply:
Hi! Friendly ping. I have updated to fix the threading issue.

> However, I'm concerned it could lead to unexpected results if ActiveRecord::Relation instances set auto_include = true since the developers intent as to how this code should behave in those main scenarios is less clear.

I agree this is not very clear, which is why we sidestepped the issue :)

Honestly, per-`ActiveRecord::Relation` is confusing to me: doing `objects.auto_include(false).to_a.first.relation_1.relation_2` would still eager load `relation_2`, which is very confusing. This is why I prefer a (threaded) block enable/disable version

Reply:
Apologies for the delay in responding. I've been out on vacation and I'm digging out from a mountain of email. I'll give this some more thought and get back to you in a few days.

Reply:
Thanks for your patience on this @fsateler. Thinking it through a bit more, a big source of my concern stems from reusing the `auto_include` term and how that interacts with `ActiveRecord::Relation#auto_include`. What do you think about something like this which will hopefully make this interaction more obvious?

```ruby
# Enable/disable Goldiloader globally across all threads
Goldiloader.enabled = false

# Enable Goldiloader for the current thread + block of code
Goldiloader.enabled do
  # ...
end

# Disable Goldiloader for the current thread + block of code
Goldiloader.disabled do
  # ...
end
```

Reply:
Sounds good to me! I'll work on that.

Reply:
Follow up questions:

1. I'm thinking the global switch should be called differently from the block switch. This is becase it should be possible to use non-block versions when you don't have a block (but still want to respect threads). For example, rspec before/after or minitest setup/teardown. What do you think?
2. What do you think goldiloader should do with relations that "escape" the enabled block?  That is: `Goldiloader.enabled{ $blogs = Blog.all } ; blogs.map(&:author)`. Should `author` be auto-included?

Reply:
1. Good point. A non-block form that only impacts the current thread would be useful (although for tests wouldn't you want to toggle the global setting so it impacts all threads?). The difference in behavior between `Goldiloader.disable!` and `Goldiloader.enable = val` is a bit subtle. I'd rather go with something that's more obvious even if it means a more verbose name. How about `Goldiloader.enabled=(val)` for the global version and then `Goldiloader.enabled_for_current_thread=(val)` with appropriate block forms for disabling/enabling in the current thread?

2. I agree that automatic eager loading should only happen if Goldiloader is enabled at the time that the association is actually traversed i.e. in the `Goldiloader.enabled{ $blogs = Blog.all } ; blogs.map(&:author)` example, `author` should not be automatically eager loaded.

Reply:
Thanks! I have pushed a new version. However, I inverted the verbosity of the options of point 1. This is because usually what you want is to affect only your thread. Newer rails versions are capable of running tests in parallel threads, and so you always want to use the thread-aware version during runtime. At initialization time you want the global version so that future threads will inherit the value, but once threads are created you don't want one thread to interfere with another.